### PR TITLE
[Offload] Introduce offload-tblgen and initial new API implementation

### PR DESCRIPTION
**Pull request by callumfare**
Introduce `offload-tblgen` and an initial implementation of a subset of the new API. The tablegen files are intended to be the single source of truth for the new API, with the header files, documentation, and others bits of source all automatically generated.

**TODO** (based on review feedback so far):
- [x] Check in the generated headers
  - [x] Add an `offload-generate` target to trigger the generation rather than building them every time
- [x] Decide how error handling should work
  - [x] Finish up new error handling implementation 
- [x] Decide naming convention
- [x] Add testing for the new API
- [x] Add tablegen specific testing
- [x] clang-tidy and use llvm:: types when possible
- [x] Add optional code location arguments
- [x] Avoid multiple returns from one function

### offload-tblgen

See the included [README](https://github.com/callumfare/llvm-project/blob/d80db06491d85444bb6f7e59d8068a22cef3a6b4/offload/new-api/API/README.md) for more information on how the API definition and generation works. I'm happy to answer any questions about it and plan to walk through it in a future LLVM Offload call.

It should be noted that struct definitions have not been fully implemented/tested as they aren't used by the initial API definitions, but finishing that off in the future shouldn't be too much work.

The tablegen tooling has been designed to be easily extended with new backends, using the classes in `RecordTypes.hpp` to abstract over the tablegen records.

### New API

Previous discussions at the LLVM/Offload meeting have brought up the need for a new API for exposing the functionality of the plugins. This change introduces a very small subset of a new API, which is primarily for testing the offload tooling and demonstrating how a new API can fit into the existing code base without being too disruptive. Exact designs for these entry points and future additions can be worked out over time.

The new API does however introduce the bare minimum functionality to implement device discovery for Unified Runtime and SYCL. This means that the `urinfo` and `sycl-ls` tools can be used on top of Offload. A (rough) implementation of a Unified Runtime adapter (aka plugin) for Offload is available [here](https://github.com/callumfare/unified-runtime/tree/offload_adapter). Our intention is to maintain this and use it to implement and test Offload API changes with SYCL.

### Demoing the new API

```sh
$ git clone -b offload_adapter https://github.com/callumfare/unified-runtime.git
$ cd unified-runtime
$ mkdir build
$ cd build
$ cmake .. -GNinja -DUR_BUILD_ADAPTER_OFFLOAD=ON \
    -DUR_OFFLOAD_INSTALL_DIR=<offload build dir containing liboffload_new.so> \
    -DUR_OFFLOAD_INCLUDE_DIR=<offload build dir containing 'offload' headers directory>
$ ninja urinfo
export LD_LIBRARY_PATH=<offload build dir containing offload plugin libraries>
$ UR_ADAPTERS_FORCE_LOAD=$PWD/lib/libur_adapter_offload.so ./bin/urinfo
[cuda:gpu][cuda:0] CUDA, NVIDIA GeForce GT 1030  [12030]
# Demo with tracing
$ OFFLOAD_TRACE=1 UR_ADAPTERS_FORCE_LOAD=$PWD/lib/libur_adapter_offload.so ./bin/urinfo
---> offloadPlatformGet(.NumEntries = 0, .phPlatforms = {}, .pNumPlatforms = 0x7ffd05e4d6e0 (2))-> OFFLOAD_RESULT_SUCCESS
---> offloadPlatformGet(.NumEntries = 2, .phPlatforms = {0x564bf4040220, 0x564bf4040240}, .pNumPlatforms = nullptr)-> OFFLOAD_RESULT_SUCCESS
...
```

### Open questions and future work
* The new API is implemented in a separate library (`liboffload_new.so`). It could just as easily be part of the existing `libomptarget` library - I have no strong feelings on which is better.
* Only some of the available device info is exposed, and not all the possible device queries needed for SYCL are implemented by the plugins. A sensible next step would be to refactor and extend the existing device info queries in the plugins. The existing info queries are all strings, but the new API introduces the ability to return any arbitrary type.
* It may be sensible at some point for the plugins to implement the new API directly, and the higher level code on top of it could be made generic, but this is more of a long-term possibility.

### CODE REVIEW

File: offload/CMakeLists.txt
Comment (jhuber6) [inline offload/CMakeLists.txt:353]: I don't think we should call this `new-api`.

File: offload/new-api/API/Common.td
Comment (jhuber6) [inline offload/new-api/API/Common.td:33]: Do we really need this? I highly doubt that we care about any compilers that don't have `stdint.h` at this point.

File: offload/new-api/API/Common.td
Comment (jhuber6) [inline offload/new-api/API/Common.td:39]: Is `platform` what we call a plugin currently and a device a subset of that? It's tough to think how we should handle this in what's exported, since I would say that ideally the user doesn't need to care about the "plugin" at all. However, that's not true for cases where some queue is only valid with CUDA and not AMDGPU for example. We could probably make it a quality of the device.

File: offload/new-api/API/Common.td
Comment (jhuber6) [inline offload/new-api/API/Common.td:49]: Bikeshedding, I think `offload_` in front of everything is too verbose.

File: offload/new-api/API/Common.td
Comment (jhuber6) [inline offload/new-api/API/Common.td:54]: I think we talked a lot about error codes, where the returned value is actually an index into some table (zero is success / invalid index). The actual error code could be queried from that index or something. This is mostly because we want the library to provide more helpful error codes than generic "X went wrong" since we already do a lot of stuff like `return Error::("Something called %s failed on device %d", name, device)` instead of just returning failure.

File: offload/new-api/CMakeLists.txt
Comment (jhuber6) [inline offload/new-api/CMakeLists.txt:14]: What is `offload_new` and what's old? What is the expected way for this to be used?

File: offload/new-api/CMakeLists.txt
Comment (jhuber6) [inline offload/new-api/CMakeLists.txt:18]: These can't be hard-coded, and why do we depend on `omptarget`? That's the OpenMP portion of the runtime while I thought this was a separate interface that merely shared the generic plugins.

**Review by jhuber6 — COMMENTED** [review]
Great, glad to see something coming along here.

Generally I'd like a standalone interface into the plugins that can be targeted without the awful OpenMP interface, this seeems to be going in that direction so I'm happy about that.

I'm guessing right now this is mostly a demo and doesn't expose a functioning kernel launch? One downside to using this TableGen stuff is I have no clue what the headers look like after being generated.

**Comment by callumfare** [comment]
@jhuber6 

Thanks for the quick review, I'll try to address your other comments asap.

> I'm guessing right now this is mostly a demo and doesn't expose a functioning kernel launch? One downside to using this TableGen stuff is I have no clue what the headers look like after being generated.

Indeed, this doesn't expose anything to do with kernels, queues, memory, etc. right now. Just basic device discovery stuff which is enough for a quick demo - see the information in the PR description. I think implementing something that exposes all that would be achievable quickly, but getting a polished interface will obviously take much longer. Either way I didn't want to stick too much in the initial PR.

I agree that not seeing the generated headers is unfortunate. In Unified Runtime we check in the generated files (and have a CI job that fails if the generated headers don't match the output of re-running the generation). All things aside I think I'd prefer that approach, but I understand that's not how any other LLVM project works so it might be difficult to get traction there.

File: offload/new-api/CMakeLists.txt
Comment (callumfare) [reply offload/new-api/CMakeLists.txt]: I've fixed this now, the `omptarget` dependency wasn't intentional and the plugins now get linked based on `LIBOMPTARGET_PLUGINS_TO_BUILD`

File: offload/new-api/API/Common.td
Comment (callumfare) [reply offload/new-api/API/Common.td:33]: This is just a copy-paste from UR and a test that the tablegen'd typedefs work. It's not actually used anywhere right now anyway and can be removed and re-added in future - whether it's really needed in practice is a good question.

File: offload/new-api/API/Common.td
Comment (callumfare) [reply offload/new-api/API/Common.td:39]: Yes, the idea of this is that `platform` is roughly equivalent to a plugin (with a big caveat*).

In SYCL, the platforms are exposed to the user (see [here](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_platform_interface)), so we can't hide these completely from the interface. Obviously that's not important to OpenMP, so anything omp related built on top of this wouldn't expose them, but in general we need something in the new API.

*In UR we actually split the concept of a plugin (called an `adapter` in UR) from the platform. This is because the OpenCL adapter can contain multiple platforms (one for every valid OpenCL driver on the system). For other adapters it's just a 1:1 relationship and not important. I didn't want to introduce this complication at this point though, especially since we don't even have an OpenCL plugin.

File: offload/new-api/API/Common.td
Comment (callumfare) [reply offload/new-api/API/Common.td:49]: Agreed, by the time I was done it was feeling a little unwieldy but I didn't want to rewrite everything without agreement - it would be good to make a decision as a group on this. `ol_` or `offl_` might be better choices

Comment (jhuber6) [reply offload/new-api/API/Common.td:49]: Also, I think there was some desire for every API function to accept some identifier information, for use with debug printing. I'm wondering if we should have some `_ident` or something suffix and `#define` the normal API entrypoints to be those with a nullptr argument.

File: offload/new-api/CMakeLists.txt
Comment (callumfare) [reply offload/new-api/CMakeLists.txt:13]: I think it makes sense for this new interface to be `offload` with the existing OpenMP interface(s) staying in `omptarget`. But since the new API is nowhere near usable I wanted the name to indicate that and didn't have any better ideas, hence `offload_new`. I don't have strong feelings on this at all so happy to go with whatever others think.

File: offload/new-api/src/offload_impl.cpp
Comment (jhuber6) [inline offload/new-api/src/offload_impl.cpp:32]: These should be LLVM small vectors, since they have small object optimizations.

File: offload/new-api/src/offload_impl.cpp
Comment (jhuber6) [inline offload/new-api/src/offload_impl.cpp:109]: We should probably codify the version right at the start.

File: offload/new-api/src/offload_impl.cpp
Comment (jhuber6) [reply offload/new-api/src/offload_impl.cpp:32]: Also, similar for anything `std::` there's probably something like `llvm::find_if` that's easier to use. Can you also run clang-tidy on the whole file to make sure everything's formatted with the right names?

**Comment by tschuett** [comment]
We have tblgen tests. Could you do something similar for Offload?
https://github.com/llvm/llvm-project/tree/main/llvm/test/TableGen

Could the tablgen emitter wrap everything into a `namespace offload` to avoid prefixes?

File: offload/new-api/src/offload_impl.cpp
Comment (tschuett) [reply offload/new-api/src/offload_impl.cpp:109]: Could the version be in the .td file?

File: offload/new-api/API/Device.td
Comment (jplehr) [inline offload/new-api/API/Device.td:38]: it's probably too early to bikeshed, but I think, function names should read more like `getOffloadDevice`.

**Review by jplehr — COMMENTED** [review]
Thank you for putting this up and moving it forward.
I looked at it briefly and will get to it later.

File: offload/new-api/API/Device.td
Comment (jhuber6) [reply offload/new-api/API/Device.td:38]: This is supposed to be a C API.

File: offload/new-api/API/Common.td
Comment (callumfare) [reply offload/new-api/API/Common.td:54]: This is definitely possible, but I have some concerns about that being the default path for returning errors. 

With SYCL we programmatically deal with error codes, converting them to appropriate SYCL exceptions etc. See [here](https://github.com/intel/llvm/blob/c22d238c417974e540beebf2c5bd6512e32926f6/sycl/source/detail/error_handling/error_handling.cpp#L353) for an example. Having errors just be strings makes this more difficult to handle, and liable to be unintentionally broken.

Could we have a specific error condition (e.g. `RESULT_ERROR_CUSTOM`) that represents a custom error state that can then be queried with a function like `offloadGetLastCustomError`? We have something similar in UR (`UR_RESULT_ERROR_ADAPTER_SPECIFIC`). Rather than using an index into a table of strings, we have a thread_local buffer for a single error message, and it's the caller's responsibility to query the contents of it after the failing API call (subsequent calls may overwrite it). This design avoids complications with the lifetime of the error strings.

Another possibility is some kind of logging callback where Offload can write out additional error/debug information for the language runtime to handle in whatever way is suitable. That would allow an enum to be used as the normal return code, while additional more descriptive information is optionally available.

Comment (callumfare) [reply offload/new-api/API/Common.td:54]: Ah ok, yeah that makes more sense, thanks!

Would the table contain one or more errors for the last API call or is the idea that the table can be queried for any previous API call that returned an error? If it's the latter then I'm still a bit unsure about how that would work in terms of lifetime; seems like they would just pile up indefinitely since you can never safely clear them. Maybe that's not a big deal in practice but it makes me a bit worried about long-running programs. That's why we made the equivalent in UR thread-local and mandated checking the error message before making another API call on the same thread, since the implementation can safely clear the existing message(s) when setting a new one.

**Comment by JonChesterfield** [comment]
Generating lots of redundant C++, where the redundancies are required by the language, is a good thing. Tablegen has pros and cons but is at least well established.

If we check in the generated code, we essentially guarantee that people will edit the generated files, and not the tablegen, and that review will miss this. The single source of truth will diverge from the thing actually running. However, if we do not check in the generated source, it'll be impossible to tell from the github review what has changed.

I would suggest we generate the source from tablegen, and check it in, and also have a unit testing step which generates the source again and diffs it against the copy in the local filesystem, and fails if they differ. As long as CI is running that test we'll catch the divergence.

Also don't call it new. This would be the... fourth? plugin revision and I think most of them are called some variant on "new".

Is it viable to develop the tablegen without the source checked in, and once it's holding together, have it replace the existing source files? Thus there is no new/old, just a transition from handwritten C++ into generated C++ which does the same thing, and we can permute the existing source code to look more like the generated output under CI.

**Comment by jhuber6** [comment]
> Also don't call it new. This would be the... fourth? plugin revision and I think most of them are called some variant on "new".

Honestly my preferred solution would just be to start this as an entirely separate interface so we don't need to deal with breaking OpenMP stuff for existing customers. That's what I wanted from `offload/` in the first place.

**Comment by callumfare** [comment]
> I would suggest we generate the source from tablegen, and check it in, and also have a unit testing step which generates the source again and diffs it against the copy in the local filesystem, and fails if they differ. As long as CI is running that test we'll catch the divergence.

Agreed, I think this is a good way to balance things. I'll make sure to add a test/check target to this PR so we can immediately have it covered by CI.

> Also don't call it new. This would be the... fourth? plugin revision and I think most of them are called some variant on "new".

I think it makes sense to just call this the Offload API while the existing stuff is the omptarget plugin interface (or similar), if there are no objections to that. Obviously in this PR the API doesn't come close to implementing a usable interface, but we can/should focus on getting the basics working (up to and including kernel execution) asap.

> Is it viable to develop the tablegen without the source checked in, and once it's holding together, have it replace the existing source files? Thus there is no new/old, just a transition from handwritten C++ into generated C++ which does the same thing, and we can permute the existing source code to look more like the generated output under CI.

Swapping it out when ready makes sense to me, although I'm not sure how we get there. I see the current approach being

- Implement the new API on top of the existing plugin interface. This is what this PR does. We can't simply change the existing plugins without breaking omptarget (or spending significant effort updating that code at the same time).
  - This also means for new functionality we will have to add it to the old interface, even in a hacky way, so the implementation has something to use. Seems like it would be extra effort on the old plugins.
- When the new API reaches some usable critical mass, update the plugins to implement the new API directly. Then either rewrite the plugin interface (that libomptarget uses) on top of that, or just rewrite the libomptarget code to use it directly. I'm not familiar enough with the project to say whether that's a good idea.
  - This is kind of an all-or-nothing thing where we can't really swap things out incrementally. 

> Honestly my preferred solution would just be to start this as an entirely separate interface so we don't need to deal with breaking OpenMP stuff for existing customers. That's what I wanted from `offload/` in the first place. i.e. redo the plugins as well.

This is absolutely our (Codeplay) preference, but I'd got the impression that we'd decided against this as a group. If we did this, we could create a `plugins-nextgen/offload` plugin (implementing the old interface on top of the new one), to allow us to optionally run OpenMP through the new API, but not breaking anything else. Then at some future point it could become the default path. I think this would be less of a headache than the approaches described above.

**Comment by jhuber6** [comment]
> This is absolutely our (Codeplay) preference, but I'd got the impression that we'd decided against this as a group.

As far as I recall, that was the preference of everyone at AMD and the only ones in favor of what we have now were Johannes and the people on the RFC that didn't contribute to `libomptarget`.

**Comment by callumfare** [comment]
> As far as I recall, that was the preference of everyone at AMD and the only ones in favor of what we have now were Johannes and the people on the RFC that didn't contribute to `libomptarget`.

I'm not against having that discussion again, before we start actually committing new code, if it makes thing easier for us. If we keep the current approach it seems important to understand how we actually transfer the existing code (both plugins and libomptarget runtime) to the new API without breaking OpenMP things.

**Comment by jhuber6** [comment]
> I'm not against having that discussion again, before we start actually committing new code, if it makes thing easier for us. If we keep the current approach it seems important to understand how we actually transfer the existing code (both plugins and libomptarget runtime) to the new API without breaking OpenMP things.

`libomptarget` is currently written in terms of the `GenericPluginTy` interface, and presumably this API can be as well. The main issue I have with that is there's a lot of OpenMP specific stuff in the plugin that prevents me from adequately fixing / modifying the interface. There's also the fact that modifying stuff upstream tends to conflict with AMD's fork (mostly our fault however).

**Comment by callumfare** [comment]
> `libomptarget` is currently written in terms of the `GenericPluginTy` interface, and presumably this API can be as well. The main issue I have with that is there's a lot of OpenMP specific stuff in the plugin that prevents me from adequately fixing / modifying the interface. There's also the fact that modifying stuff upstream tends to conflict with AMD's fork (mostly our fault however).

Ok, I was getting confused between the `GenericPluginTy` interface (which we can presumably change as much as is needed to enable the new API) and the plugin-specific bits of libomptarget (`__tgt_rtl_*`) which I'm assuming we can't change without breaking downstream projects.

**Comment by jhuber6** [comment]
> Ok, I was getting confused between the `GenericPluginTy` interface (which we can presumably change as much as is needed to enable the new API) and the plugin-specific bits of libomptarget (`__tgt_rtl_*`) which I'm assuming we can't change without breaking downstream projects.

The `__tgt_rtl_*` functions used to be a pseudo-API because the plugins were all `.so` files loaded via `dlopen`. I reworked that but never go around to modifying the functions themselves.

**Comment by callumfare** [comment]
I've added unit tests. The `LibomptUnitTests` target now builds a `offload.unittests` executable. It optionally takes a `--platform` argument to specify the platform/plugin to use, and otherwise picks a sensible default (i.e. one with an available device).

In general it should be possible to fully parameterize the tests for every available platform/device but I've not gone that far at this stage - when the API is a little more fully formed I'm happy to go back and do that.

Incidentally, I disabled the existing `NextgenPluginsTest.cpp` unit test because it didn't even build and seems like it targets an older version of the plugin interface. Probably could just be deleted?

File: offload/new-api/API/Common.td
Comment (jhuber6) [reply offload/new-api/API/Common.td:54]: We could also do pointers or something, null is no error. That might make it easier to implement the "table" as a thread safe data structure (Since otherwise an index could never be reused).

**Comment by callumfare** [comment]
I've checked in the autogenerated files. The main issue with them now is that they're not clang-format clean. I think trying to maintain the tablegen backends so they always generate clang-format clean files isn't realistic, so I'd like to include a dependency on clang-format so they can be formatted after being generated. It's not clear to me how to achieve this with a runtime build when the clang-format tool is part of the parent LLVM project, but I'm most likely missing something obvious. Relying on whatever clang-format version is installed on the system would mean inconsistent results (unless we mandate a specific version).

If anyone has any ideas on achieving this suggestions are very welcome. Or alternatively this might be a reason to decide against checking in all the generated files.

**Comment by jhuber6** [comment]
> I've checked in the autogenerated files. The main issue with them now is that they're not clang-format clean. I think trying to maintain the tablegen backends so they always generate clang-format clean files isn't realistic, so I'd like to include a dependency on clang-format so they can be formatted after being generated. It's not clear to me how to achieve this with a runtime build when the clang-format tool is part of the parent LLVM project, but I'm most likely missing something obvious. Relying on whatever clang-format version is installed on the system would mean inconsistent results (unless we mandate a specific version).
> 
> If anyone has any ideas on achieving this suggestions are very welcome. Or alternatively this might be a reason to decide against checking in all the generated files.

You can just pull it out of the LLVM binary path with `find_program(clang-format clang PATHS ${LLVM_TOOLS_BINARY_DIR} NO_DEFAULT_PATH)` or similar. Failing that probably shouldn't be a hard error.

**Comment by callumfare** [comment]
Thanks @jhuber6, I have it all working now. The depfile issue was some weird behavior with relative paths so I just moved where the relevant cmake around a bit.

In summary:
* The generated files are all checked in and formatted, no tablegen is required to build the library.
* If someone changes the API, they need to build the `OffloadGenerate` target to regenerate the files which are all copied back to the in-tree directory. If clang-format isn't found, the target won't be available.

**Comment by tschuett** [comment]
I don't understand why you check in auto-generated code. It is a common anti-pattern. You should instead make tblgen a build dependency. It is the common approach in the rest of the project. If you run e.g. `ninja llc`, there will  be lot of tblgen invocations if the output was not generated in a previous run.

**Comment by tschuett** [comment]
We have `ninja install-llvm-headers` for a reason. External applications should not build on a mono repo checkout. You have to install llvm before using it.

**Comment by jhuber6** [comment]
> I don't understand why you check in auto-generated code. It is a common anti-pattern. You should instead make tblgen a build dependency. It is the common approach in the rest of the project. If you run e.g. `ninja llc`, there will be lot of tblgen invocations if the output was not generated in a previous run.

The main concern here is just that it's really difficult to review changes since we don't see the header (Since this is an API header the user will read and not just some implementation detail). I can see it either way honestly, it's just tough to review without seeing the changes.

**Comment by tschuett** [comment]
> > I don't understand why you check in auto-generated code. It is a common anti-pattern. You should instead make tblgen a build dependency. It is the common approach in the rest of the project. If you run e.g. `ninja llc`, there will be lot of tblgen invocations if the output was not generated in a previous run.
> 
> The main concern here is just that it's really difficult to review changes since we don't see the header (Since this is an API header the user will read and not just some implementation detail). I can see it either way honestly, it's just tough to review without seeing the changes.

Users will do the same thing as for llvm: `ninja install-offload-headers install-offload-libraries`.

When I invoked `ninja install-llvm-headers`, it first started to build support. The installed llvm headers depend on tblgen.

**Comment by callumfare** [comment]
One other reason to check in the headers is so that we can rely on them being clang-formatted without having a hard dependency on clang-format. I assume this is better for standalone builds and other unusual configurations.

Perhaps as a compromise the user facing headers (`offload_api.h`, and probably `offload_print.hpp`) could be checked in but the other files (which are purely implementation details) left to be auto-generated normally, since we don't particularly care about them being formatted correctly.

I think the reviewability will be less of an issue after this initial PR since future changes will be smaller and reviewers will be more familiar with what gets generated. I still have a preference for checking in the files but I understand it's not how tablegen has been used so far so it's controversial.

**Comment by tschuett** [comment]
* On your machine is probably a file called arm_sve.h. Surprise it is tablgened, but easy to read.
* If you are eager to checkin auto-generated files, then maybe tblgen is not the right tool for you.
* Maybe you have to improve your tblgen tool, to make the output look nicer.

**Comment by callumfare** [comment]
@jhuber6  I've reworked the implementation of errors based on the discussion in the last call. I'm away for the next 2 weeks and haven't had time to polish it and update the UR adapter so I've not added it to this branch/PR, but you can see the changes on my `offload_new_api_error_rework` branch here:
https://github.com/llvm/llvm-project/commit/1b53217ba4703aeae071a3056890e7ab952afe02 You can use the unit tests to play around with it.

Things are identical in terms of how implementation functions return errors, but the entry points now return a pointer to a error information struct, containing an error code and optionally an error detail string. Success is just `NULL`.
The WIP elements are:
* Printing is a bit janky but should work
* Need a better container type for the error structs to avoid duplication of repeated errors. The error strings are de-duplicated in a `llvm::StringSet`, but the error structs are just allocated and put in a vector. set/unordered_set aren't suitable because of invalidation guarantees and strict weak ordering requirements. I think it's doable with a wrapper over the error allocations, so no big deal, but I've run out of time

File: offload/new-api/include/offload_api.h
Comment (jhuber6) [inline offload/new-api/include/offload_api.h:361]: I think the very first thing we should have implemented is a proper reference counted `init/shutdown` function pair. We probably not need to initialize plugins until the user tries to get a device though.

File: offload/new-api/API/Common.td
Comment (callumfare) [reply offload/new-api/API/Common.td]: I've now implemented the errors as discussed in previous calls. Success is a nullptr, an error is a pointer to a struct with an error code and an optional string.

File: offload/new-api/src/offload_impl.cpp
Comment (callumfare) [reply offload/new-api/src/offload_impl.cpp]: I've added version information to the tablegen directly. I've just set it to 0.0.1 for now. As we approach a stable-ish API we can decide how to handle versioning, whether it's just tied to the LLVM version etc.

File: offload/new-api/include/offload_api.h
Comment (callumfare) [reply offload/new-api/include/offload_api.h]: I've added `offloadInit` (which initializes the plugins) and `offloadShutDown` (which does nothing at the moment)

File: offload/new-api/include/offload_exports
Comment (jhuber6) [inline offload/new-api/include/offload_exports:2]: We should just put `ol_` as the prefix and then use `ol_*`

File: offload/unittests/OffloadAPI/device/offloadDeviceGet.cpp
Comment (jhuber6) [inline offload/unittests/OffloadAPI/device/offloadDeviceGet.cpp:31]: nit, LLVM style shouldn't put braces around single lines.
```suggestion
  if (Count < 2)
    GTEST_SKIP() << "Only one device is available on this platform.";
```

**Review by jhuber6 — COMMENTED** [review]
Generally I'm in favor of `ol_` as a prefix, `snake_case` for the C API, and `get_x` as the verbage.

**Comment by callumfare** [comment]
> Generally I'm in favor of `ol_` as a prefix, `snake_case` for the C API, and `get_x` as the verbage.

I'll update the prefix etc. Just wondering though do we want to prefer `olCamelCase` instead to better match the LLVM style rather than just do what omp does? IMO snake_case looks odd unless function parameters also use snake_case, and at that point it kind of pollutes the implementation code with that style

**Comment by jhuber6** [comment]
> > Generally I'm in favor of `ol_` as a prefix, `snake_case` for the C API, and `get_x` as the verbage.
> 
> I'll update the prefix etc. Just wondering though do we want to prefer `olCamelCase` instead to better match the LLVM style rather than just do what omp does? IMO snake_case looks odd unless function parameters also use snake_case, and at that point it kind of pollutes the implementation code with that style

Tough to say, pretty much up to personal preference. We could probably make the argument for camelCase because it's used in CUDA, Vulkan, OpenCL, and OpenGL while AFAIK `snake_case` is only in HSA.

File: offload/unittests/OffloadAPI/common/Environment.cpp
Comment (callumfare) [reply offload/unittests/OffloadAPI/common/Environment.cpp:78]: This is in the `offload.unittest` executable rather than the runtime

Comment (jhuber6) [reply offload/unittests/OffloadAPI/common/Environment.cpp:78]: Duh, didn't see any test checks and didn't parse enough of the filename.

**Review by jhuber6 — APPROVED** [review]
LG for now, we can adjust as we move forward.

**Comment by shiltian** [comment]
I'm afraid the code style of this entire stuff will be out of control. Looking at the code, there are at least three different code styles. I don't think it's a good idea to keep as it is.

**Comment by jhuber6** [comment]
> I'm afraid the code style of this entire stuff will be out of control. Looking at the code, there are at least three different code styles. I don't think it's a good idea to keep as it is.

The public API and internal style can be different like with `libomp`, but it would definitely help to be consistent.

> > we can adjust as we move forward.
> 
> IMHO, most of the time with this, it roughly means, this is it, and eventually we get so many half-baked stuff.

To be fair I've deleted some of the half baked stuff in the past, but I get your point. Generally I'm less concerned overall since I've somewhat resigned myself to whatever the offload project is going to be. This one as-is can't do anything useful, but it's hard to track changes with such a big delta. We could probably make the argument that this could be opt-in or something.

**Comment by callumfare** [comment]
Thanks for the review @jhuber6!


>> I'm afraid the code style of this entire stuff will be out of control. Looking at the code, there are at least three different code styles. I don't think it's a good idea to keep as it is.
>
> The public API and internal style can be different like with libomp, but it would definitely help to be consistent.

I've used clang-tidy to keep things as consistent as possible but there are two distinct styles:
- In the public API: `ol_snake_case_t` for types, `olCamelCase` for functions
- In the implementation: Normal LLVM style.

I've kept the public API in that style since it's in line with most other C APIs. If anything the previous conversation was leaning towards making everything snake_case.

If there's any inconsistency beyond that then that's a mistake and I'm happy to correct it.

@shiltian We've previously discussed various aspects of style at the bi-weekly LLVM Offload calls and I think what we have now is as close to a consensus as we could get. I'm not against changing the style again to be purely LLVM style, but I'd want to discuss it as a group again at the next call (December 11th). If you'd like to join that call I'd be happy to have that conversation. In the meantime I would strongly prefer merging this as-is so progress isn't held up. The tablegen tooling means any style or design changes in the future should be fairly painless.

> IMHO, most of the time with this, it roughly means, this is it, and eventually we get so many half-baked stuff.

I understand the concern, but we have many more contributions planned after this. I didn't want the scope of this initial PR to get out of control so have avoided implementing too much of the API initially. In the very short term we want to extend the API to expose all the existing functionality in the libomptarget plugins. Beyond that we want to extend functionality to enable SYCL on top of Offload.

We (Codeplay) plan on having multiple engineers working on this, once this PR is in we should be able to start ramping up effort. We've been attending the bi-weekly Offload meetings and discussing our plans there, and will continue to do so. We're happy to discuss any specific details during a future call.

@jhuber6 I don't have write access. If you're still happy with everything after the above comments would you mind landing the change?

**Comment by callumfare** [comment]
@jhuber6 Gentle nudge on the above

**Comment by jhuber6** [comment]
@callumfare ^

**Pull request merged**
